### DISCUSSION TITLE
[COMPOSITE] Icones Normal/Défaut avec Couleurs

### DISCUSSION THREAD

**Discussion by WITNicolas** [witsa/synapps#500] [Show and tell]
Un Composite qui permet de choisir 2 logos (Normal/Défaut) ainsi que 2 couleurs (Normal/Defaut).
En mode Défaut, le logo clignote.
Il suffit de liaisonner la variable etatDefaut

![defClignotant](https://user-images.githubusercontent.com/33632664/171877744-4611ceb4-8052-48c3-8d8f-f8fa411c5734.gif)

A coller dans les composites.

````text
SYNAPPS-STUDIO-COMPOSITE|{"config":{"key":"defClignotant","name":"defClignotant","properties":{"fontSize":"2em"},"additionalDefs":{"iconNOK":{"type":"icon","canBeEmpty":false,"value":"danger"},"iconOK":{"type":"icon","value":"circleselect"},"colorNOK":{"type":"color","helperTxt":"Couleur Etat Défaut","value":"red"},"colorOK":{"type":"color","helperTxt":"Couleur Etat Normal","value":"green"},"etatDefaut":{"type":"boolean","description":"Lien du Défaut","value":false}}},"leadActor":{"type":"layout/stack","key":"stackRoot","children":[{"type":"display/html","key":"def","properties":{"content":"<style>\n@keyframes condemed_blink_effect {\n  from {\n    opacity: 0.2;\n  }\n  to {\n    opacity: 0.8;\n  }\n}\n\n/* .actor.blink {\n  animation: 500ms linear infinite alternate condemed_blink_effect;\n} */\n</style>\n\n<i class=\"icon-{{icon}}\"></i>","color":"red","top":"470px","left":"1080px"},"additionalDefs":{"icon":{"type":"icon","label":"Icône","description":"Web Hosting Hub Glyphs"}},"additionals":{"icon":"danger"},"bindings":{"additionals.icon":"stage@properties.etatDefaut","properties.color":"stage@properties.etatDefaut","properties.animation":"stage@properties.etatDefaut"},"events":{"additionals/icon/binding/onReadTransform":["return context.value ? this.stage.properties.iconNOK : this.stage.properties.iconOK;"],"properties/color/binding/onReadTransform":["return context.value ? this.stage.properties.colorNOK : this.stage.properties.colorOK;"],"properties/animation/binding/onReadTransform":["return context.value ? '500ms linear infinite alternate condemed_blink_effect' : '';"]}}]}}

**Comment by edejean** [comment]
Bonjour Nicolas,
Et si je veux la même chose sur le fond, un cadre vert légèrement transparent entourant par exemple un disjoncteur et passant au rouge une fois déclenchée ? pas besoin de l'îcone...

**Reply by manufitoussiwit** [reply]
Salut, il existe plusieurs moyens d'obtenir ce résultat.
Je peux proposer ceci : 

![demo style default2](https://user-images.githubusercontent.com/35595723/182633155-7acfff52-733e-47e9-95f5-7f648e2aaa9f.gif)


```text
SYNAPPS-STUDIO-COMPOSITE|{"config":{"key":"bg-default","name":"BG Defaut","additionalDefs":{"isDefault":{"type":"boolean","label":"En défaut?","value":false},"targetKey":{"type":"actor","label":"Acteur ciblé","value":""},"rc":{"type":"color","label":"Couleur normale","value":"rgba(37, 207, 48, 0.42)"},"c":{"type":"color","label":"Couleur en défaut","value":"rgba(198, 77, 77, 0.68)"},"duration":{"type":"number","label":"Durée clignotement","value":600,"min":0,"step":100,"unit":"ms"}}},"leadActor":{"type":"layout/stack","key":"stack1","children":[{"type":"display/html","key":"bg-default-style","properties":{"content":"<style>\n  @keyframes {{uniqKey}}_is_in_default {\n    from {\n      background-color: none;\n    }\n\n    to {\n      background-color: {{c}};\n    }\n  }\n</style>"},"additionalDefs":{"c":{"type":"color","label":"Couleur en défaut"},"isDefault":{"type":"boolean","label":"En défaut?"},"rc":{"type":"color","label":"Couleur normale"},"targetKey":{"type":"actor","label":"Acteur ciblé"},"duration":{"type":"number","label":"Durée clignotement","min":0,"step":100,"unit":"ms"},"uniqKey":{"type":"text"}},"additionals":{"c":"rgba(198, 77, 77, 0.68)","rc":"rgba(37, 207, 48, 0.42)","duration":600},"bindings":{"additionals.c":"stage@properties.c","additionals.targetKey":"stage@properties.targetKey","additionals.isDefault":"stage@properties.isDefault","additionals.rc":"stage@properties.rc","additionals.duration":"stage@properties.duration"},"events":{"onPropertyChanged":["","if ([","  'additionals.targetKey',","  'additionals.duration',","  'additionals.isDefault',","  'additionals.rc',","].includes(context.propertyPath)) {","","  if (context.propertyPath === 'additionals.targetKey') {","    const oldTargetKey = context.oldValue;","    if (!context.utils.isEmpty(oldTargetKey)) {","      const oldTarget = context.stage.stage.getActor(oldTargetKey);","      if (oldTarget) {","        oldTarget.properties.reset('animation');","        oldTarget.properties.reset('backgroundColor');","      }","    }","  }","","  this.runEvent('onPostInit');","}"],"onPostInit":["const targetKey = this.additionals.targetKey;","if (context.utils.isEmpty(targetKey)) return;","const target = context.stage.stage.getActor(targetKey);","if (context.utils.isNone(target)) return;","this.additionals.uniqKey = `${target.uniqClass}_${this.stage.uniqClass}`;","if (this.additionals.isDefault) {","  target.properties.animation = `${this.additionals.duration}ms linear infinite alternate ${this.additionals.uniqKey}_is_in_default`;","  target.properties.reset('backgroundColor');","} else {","  target.properties.reset('animation');","  target.properties.backgroundColor = this.additionals.rc;","}"]}}]}}
```

C'est un composite à ajouter dans le projet (à copier/coller dans les composites). Ensuite il faut en ajouter une instance dans la scène, pas loin de celui qu'on veut faire clignoter. 

Dans ces propriétés on doit choisir l'acteur ciblé. 

Ensuite on peut paramétrer la couleur normale, la couleur défaut, la durée de clignotement. 

Et bien sur, il y une propriété qui permet d'indiquer s'il est en défaut ou pas, à lier à la donnée.

**Reply by edejean** [reply]
Nickel !!!!
Juste une petite remarque, dans la mesure où on utilise 2 acteurs pour arriver à cet effet, ça aurait été sympa qu'on puisse également appliquer cet effet à plusieurs acteurs "cible" depuis ce composite. (j'ai une même DI qui anime plusieurs acteurs, je dois dupliquer la paire autant de fois que nécessaire...)
Cdt
Eric

**Reply by MaximeRoby** [reply]
Hello,
j'ai beau lié la donnée, quand je la change d'état, elle reste toujours en défaut même sur off. Il y a un truc qui m'échappe?

**Reply by MaximeRoby** [reply]
Autant pour moi, soucis avec le rafraichissement!

**Comment by WITNicolas** [comment]
Nouvelle version du composite.
La derniere mise à jour de STUDIO à mis en évidence une erreur de conception de base du composite, qui n'aurait jamais dû fonctionner en fait.....
Vous pouvez supprimer l'ancien composite et coller cette nouvelle version.

**Reply by manufitoussiwit** [reply]
Je l'ai mis à jour aussi dans la section des composites remarquables https://witsa.github.io/synapps/concepts/composite.html#ic%C3%B4ne-clignotant

**Reply by WitDavid1** [reply]
Bonjour,
Comment fait t'on pour liée notre composite défaut 2 logo a notre ressources défaut du redy?
Merci

**Comment by WitDavid1** [comment]
Bonjour,
Problème avec le composite défaut clignotant, je les liée a la ressources défaut mais il ne change pas d'état?
Voir capture d'écran des liens.
![Capture d’écran 2025-04-07 073136](https://github.com/user-attachments/assets/1045dfa8-e031-4473-9280-8435ab87c801)
Merci pour vos retour.

**Reply by WITNicolas** [reply]
Tu as liaisonné l'ETAT de la ressource, il faut prendre le .OUTPUT avec le champ VALEUR.
Et seulement en LECTURE .... pas d'ECRITURE

**Comment by WitDavid1** [comment]
Bonjour,
Ok merci
J'essaie cela demain matin.


________________________________
